### PR TITLE
Obmc

### LINKED ISSUE #677: OBMC motion mode
trello reference:   https://trello.com/c/XFV6W81A/22-obmc
Working branch: https://github.com/OpenVisualCloud/SVT-AV1/tree/OBMC

### PR DESCRIPTION
**Pull request by okhlif**
**Description**
OBMC motion mode

Closes #677 

**Author**
@chkngit 
@okhlif 

**Type of change**
New feature/Tool

**Tests and performance**
BDR gain is 0.57% [All objective-fast1 clips ]
8% memory increase for 360p
Speed cost is about 11% [4 core machine]

### CODE REVIEW

File: Source/Lib/Common/Codec/EbModeDecisionProcess.c [hunk 1]
```diff
@@ -477,7 +477,30 @@ void reset_mode_decision(
         && !(frm_hdr->frame_type == KEY_FRAME || frm_hdr->frame_type == INTRA_ONLY_FRAME)
         && !frm_hdr->error_resilient_mode;
     frm_hdr->is_motion_mode_switchable = frm_hdr->allow_warped_motion;
+#if OBMC_FLAG
```
Comment (chkngit) [inline Source/Lib/Common/Codec/EbModeDecisionProcess.c]: did you include that bug fix  RESET_BUG_FIX here?

File: Source/Lib/Common/Codec/EbProductCodingLoop.c [hunk 2]
```diff
@@ -1644,6 +1658,9 @@ void set_md_stage_counts(
 
 #if II_COMP_FLAG
     context_ptr->md_stage_2_count[CAND_CLASS_4] = context_ptr->bypass_stage1[CAND_CLASS_4] ? context_ptr->md_stage_1_count[CAND_CLASS_4] : (picture_control_set_ptr->slice_type == I_SLICE) ? 0 : (picture_control_set_ptr->parent_pcs_ptr->is_used_as_reference_flag) ? 12: 4;// 14 : 4;
```
Comment (chkngit) [inline Source/Lib/Common/Codec/EbProductCodingLoop.c]: to remove the check on bypass flag here

File: Source/Lib/Common/Codec/EbModeDecisionProcess.c (hunk 1 above)
Comment (okhlif) [reply Source/Lib/Common/Codec/EbModeDecisionProcess.c]: Done Thank you

File: Source/Lib/Common/Codec/EbProductCodingLoop.c (hunk 2 above)
Comment (okhlif) [reply Source/Lib/Common/Codec/EbProductCodingLoop.c]: Done Thank you

File: Source/Lib/Common/Codec/EbDefinitions.h [hunk 3]
```diff
@@ -33,6 +33,11 @@
 extern "C" {
 #endif
 
+#define OBMC_FLAG            1 // OBMC motion mode flag
+#define OBMC_CONVOLVE        1 // to track convolve kernels changes
+
+#define RESET_BUG_FIX        1
```
Comment (hassount) [inline Source/Lib/Common/Codec/EbDefinitions.h]: Can we merge this macro with the one of the other OBMC marcos or add a clear description of what the fix is ?

File: Source/Lib/Common/Codec/EbEntropyCoding.c [hunk 4]
```diff
@@ -1604,6 +1604,52 @@ static INLINE int is_global_mv_block(
...
+
+    if (is_motion_variation_allowed_bsize(bsize) &&
+        is_inter_mode(mode) &&
+        rf1 != INTRA_FRAME &&
+        !(rf1 > INTRA_FRAME)) // is_motion_variation_allowed_compound
+    {
+        if (!has_overlappable_candidates(cu_ptr)) // check_num_overlappable_neighbors
+            return SIMPLE_TRANSLATION;
+
+        // if (cu_ptr->prediction_unit_array[0].num_proj_ref >= 1 &&
```
Comment (hassount) [inline Source/Lib/Common/Codec/EbEntropyCoding.c]: is this code needed ?

**Comment by hassount** [comment]
@okhlif 
what is the memory impact of this feature ?

File: Source/Lib/Common/Codec/EbDefinitions.h (hunk 3 above)
Comment (okhlif) [reply Source/Lib/Common/Codec/EbDefinitions.h]: Added a comment, Thank you

**Comment by okhlif** [comment]
@hassount  8% memory increase for 360p

File: Source/Lib/Common/Codec/EbEntropyCoding.c (hunk 4 above)
Comment (okhlif) [reply Source/Lib/Common/Codec/EbEntropyCoding.c]: Removed thank you

**Comment by joelsole** [comment]
Adding OBMC to an encoder is a big effort. My concern with PR is that there is a lot of code duplication: OBMC is already present in the decoder and its code should be reused as much as possible. Example of duplications:
- There is already the file /Lib/Common/ASM_AVX2/EbBlend_a64_mask_avx2.c - no need to add blend_a64_mask_sse4.c. 
- Functions already defined at the decoder side are done again, e.g., eb_av1_skip_u4x4_pred_in_obmc (named av1_skip_u4x4_pred_in_obmc), av1_get_obmc_mask, eb_aom_highbd_blend_a64_vmask_c, etc...
- Functions like foreach_overlappable_nb_above or foreach_overlappable_nb_left are implemented in SVT-AV1 in a simplified way with: count_overlappable_nb_above, count_overlappable_nb_left, eb_av1_count_overlappable_neighbors. Probably, there is not need to bring in all that circuitry from libaom.
- Are two similar structures (build_prediction_ctxt and build_prediction_hbd_ctxt) needed? They are almost the same - could they be combined?

Could you re-use as much code as possible, bringing the necessary code from the decoder to common? Renaming them would be also helpful, as it is confusing to find functions prefixed 'dec_' in the encoder.

There is issue 683 (https://github.com/OpenVisualCloud/SVT-AV1/issues/683) on this harmonization. We should strive to harmonize encoder and decoder as much as possible.

**Comment by chkngit** [comment]
@joelsole  I second all your comments. OBMC is one of last big features that will land in the code along with Palette + global motion (hopefully in the next few weeks). we will definitely need to go aver a big code refactoring project involving enc/dec harmonization 8/10bit harmonization and code reformatting (with the new clang format).   Addressing the points you brought up for obmc make sense.

**Comment by okhlif** [comment]
@joelsole  I agree totally. Please have a look. I removed redundant functions and merged a structure. I also tried to harmonize functions between decoder and encoder. 
I will keep a eye on previous features to do the same in future PRs.

**Comment by chkngit** [comment]
will be good to get this PR in as other PRs will be based on this one

File: Source/Lib/Common/ASM_SSE4_1/EbBlend_a64_mask_sse4.c [hunk 5]
```diff
@@ -1203,3 +1203,177 @@ void aom_highbd_blend_a64_hmask_sse4_1(
         src1_8, src1_stride, mask, 0, w, h, 0, 0,
         bd);
 }
+void eb_aom_highbd_blend_a64_mask_c(uint16_t *dst, uint32_t dst_stride,
+                                 const uint16_t *src0, uint32_t src0_stride,
```
Comment (joelsole) [inline Source/Lib/Common/ASM_SSE4_1/EbBlend_a64_mask_sse4.c]: can you reuse the existing aom_highbd_blend_a64_mask_c? The casting to 16 bit is done in the function.

Comment (okhlif) [reply Source/Lib/Common/ASM_SSE4_1/EbBlend_a64_mask_sse4.c]: Thank you, Addressed.

File: Source/Lib/Common/Codec/EbInterPrediction.c [hunk 6]
```diff
@@ -2746,187 +2756,1838 @@ void search_compound_avg_dist(
...
+    int32_t subpel_x, subpel_y;
+    uint16_t * src_ptr;
+    uint16_t * dst_ptr;
+    int32_t src_stride;
+    int32_t dst_stride;
+    ConvolveParams conv_params;
+    InterpFilterParams filter_params_x, filter_params_y;
+
+
+    {
```
Comment (mafonsoN) [inline Source/Lib/Common/Codec/EbInterPrediction.c]: extra new lines here

File: Source/Lib/Common/Codec/EbInterPrediction.c [hunk 7]
```diff
@@ -2746,187 +2756,1838 @@ void search_compound_avg_dist(
...
+            subpel_x,
+            subpel_y,
+            &conv_params,
+            10);
+
+    }
+
+
+
+    return return_error;
```
Comment (mafonsoN) [inline Source/Lib/Common/Codec/EbInterPrediction.c]: Extra new lines

File: Source/Lib/Common/Codec/EbInterPrediction.c [hunk 8]
```diff
@@ -2746,187 +2756,1838 @@ void search_compound_avg_dist(
...
+            mv_q4 = clamp_mv_to_umv_border_sb(xd, &mv, bwidth, bheight, 1, 1);
+            subpel_x = mv_q4.col & SUBPEL_MASK;
+            subpel_y = mv_q4.row & SUBPEL_MASK;
+            src_ptr = src_ptr + (mv_q4.row >> SUBPEL_BITS) * src_stride + (mv_q4.col >> SUBPEL_BITS);
+            conv_params = get_conv_params_no_round(0, 0, 0, tmp_dstCr, 64, is_compound, EB_10BIT);
+            av1_get_convolve_filter_params(interp_filters, &filter_params_x,
+                &filter_params_y, bwidth, bheight);
+
+
+            convolveHbd[subpel_x != 0][subpel_y != 0][is_compound](
```
Comment (mafonsoN) [inline Source/Lib/Common/Codec/EbInterPrediction.c]: extra new lines

File: Source/Lib/Common/Codec/EbInterPrediction.c [hunk 9]
```diff
@@ -2746,187 +2756,1838 @@ void search_compound_avg_dist(
...
+            &filter_params_y,
+            subpel_x,
+            subpel_y,
+            &conv_params);
+
+    }
+
+
+
+    return return_error;
```
Comment (mafonsoN) [inline Source/Lib/Common/Codec/EbInterPrediction.c]: extra new lines. Several examples.

File: Source/Lib/Common/Codec/EbInterPrediction.c [hunk 10]
```diff
@@ -2746,187 +2756,1838 @@ void search_compound_avg_dist(
...
+
+void eb_aom_highbd_blend_a64_hmask_c(uint16_t *dst, uint32_t dst_stride,
+                                  const uint16_t *src0, uint32_t src0_stride,
+                                  const uint16_t *src1, uint32_t src1_stride,
+                                  const uint8_t *mask, int w, int h, int bd) {
+  int i, j;
+  //uint16_t *dst = CONVERT_TO_SHORTPTR(dst_8);
+  //const uint16_t *src0 = CONVERT_TO_SHORTPTR(src0_8);
+  //const uint16_t *src1 = CONVERT_TO_SHORTPTR(src1_8);
+  (void)bd;
```
Comment (mafonsoN) [inline Source/Lib/Common/Codec/EbInterPrediction.c]: remove comments as not needed

File: Source/Lib/Common/Codec/EbInterPrediction.c [hunk 11]
```diff
@@ -2746,187 +2756,1838 @@ void search_compound_avg_dist(
...
+  assert(bd == 8 || bd == 10 || bd == 12);
+
+  for (i = 0; i < h; ++i) {
+    for (j = 0; j < w; ++j) {
+      dst[i * dst_stride + j] = AOM_BLEND_A64(
+          mask[j], src0[i * src0_stride + j], src1[i * src1_stride + j]);
+    }
+  }
+}
+void eb_aom_highbd_blend_a64_vmask_c(uint16_t *dst, uint32_t dst_stride,
```
Comment (mafonsoN) [inline Source/Lib/Common/Codec/EbInterPrediction.c]: no new line between functions

File: Source/Lib/Common/Codec/EbInterPrediction.c [hunk 12]
```diff
@@ -2746,187 +2756,1838 @@ void search_compound_avg_dist(
...
+
+
+
+static INLINE void build_obmc_inter_pred_above_hbd(
+    uint8_t is16bit ,MacroBlockD *xd, int rel_mi_col,
+    uint8_t above_mi_width,
+    MbModeInfo *above_mi,
+    void *fun_ctxt,
+    const int num_planes)
+{
```
Comment (mafonsoN) [inline Source/Lib/Common/Codec/EbInterPrediction.c]: Location of the curly braces in the function definition is not consistent

File: Source/Lib/Common/Codec/EbDefinitions.h [hunk 13]
```diff
@@ -133,7 +138,11 @@ enum {
 #define BLOCK_MAX_COUNT_SB_64                     1101  // TODO: reduce alloction for 64x64
 #define MAX_TXB_COUNT                             4 // Maximum number of transform blocks.
 #if II_COMP_FLAG
+#if OBMC_FLAG
+#define MAX_NFL                                 105
```
Comment (mafonsoN) [inline Source/Lib/Common/Codec/EbDefinitions.h]: Please add a comment here to define what this is

Comment (okhlif) [reply Source/Lib/Common/Codec/EbDefinitions.h]: Added Thank you

File: Source/Lib/Common/Codec/EbInterPrediction.c (hunk 12 above)
Comment (okhlif) [reply Source/Lib/Common/Codec/EbInterPrediction.c]: Thank you, Went over the code and fixed two of them, please let me know if you find others

File: Source/Lib/Common/Codec/EbInterPrediction.c (hunk 10 above)
Comment (okhlif) [reply Source/Lib/Common/Codec/EbInterPrediction.c]: Removed, Thank you

File: Source/Lib/Common/Codec/EbInterPrediction.c (hunk 11 above)
Comment (okhlif) [reply Source/Lib/Common/Codec/EbInterPrediction.c]: Added Thank you

File: Source/Lib/Common/Codec/EbInterPrediction.c (hunk 7 above)
Comment (okhlif) [reply Source/Lib/Common/Codec/EbInterPrediction.c]: Removed few of them , Thank you

File: Source/Lib/Common/Codec/EbInterPrediction.c (hunk 8 above)
Comment (okhlif) [reply Source/Lib/Common/Codec/EbInterPrediction.c]: Removed, Thank you

File: Source/Lib/Common/Codec/EbInterPrediction.c (hunk 9 above)
Comment (okhlif) [reply Source/Lib/Common/Codec/EbInterPrediction.c]: Removed Thank you

File: Source/Lib/Common/Codec/EbInterPrediction.c (hunk 6 above)
Comment (okhlif) [reply Source/Lib/Common/Codec/EbInterPrediction.c]: removed Thank you

**Pull request merged**